java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -770,22 +802,22 @@ class DatabaseImpl(
                 arrayOf(PrimitiveType.Text.ordinal.toString())
             )
 
-            // Now delete collection_entries for those fields we just cleared.
-            val usedFieldCollectionIdsQuery =
-                """
-                    SELECT collection_id
-                    FROM fields
-                    LEFT JOIN field_values
-                        ON field_values.field_id = fields.id
-                    LEFT JOIN collection_entries
-                        ON fields.is_collection = 1
-                        AND collection_entries.collection_id = field_values.value_id
-                    WHERE fields.is_collection = 1
-                """.trimIndent()
+            // Remove all references to these entities.
+            // TODO: persist the entity storage key in this table instead of instead of concating
+            // backing key and entity id.
+            delete(
+                TABLE_ENTITY_REFS,
+                "backing_storage_key||\"/\"||entity_id in (?)",
+                arrayOf(storageKeyIds.map { it.second }.joinToString())
+            )
+            // Remove from collection_entries all references to the expired entities.            
             delete(
                 TABLE_COLLECTION_ENTRIES,
-                "collection_id NOT IN ($usedFieldCollectionIdsQuery)",
-                arrayOf()
+                """
+                    collection_id IN (SELECT id FROM collections WHERE type_id >= ?)
+                    AND value_id NOT IN (SELECT id FROM entity_refs)
+                """.trimIndent(),
+                arrayOf(PrimitiveType.values().size.toString()) // only entity collections.
```

Comment:
Might be nice to define a helper constant on PrimitiveType like PrimitiveType.LARGEST_PRIMITIVE_TYPE_ID or something similar. Could use it in isPrimitiveType too